src/mstr.c

```diff
@@ -0,0 +1,518 @@
+/*
+ * Copyright Redis Ltd. 2024 - present
+ *
+ * Licensed under your choice of the Redis Source Available License 2.0 (RSALv2)
+ * or the Server Side Public License v1 (SSPLv1).
+ */
+
+#include <string.h>
+#include <assert.h>
+#include "sdsalloc.h"
+#include "mstr.h"
+#include "stdio.h"
+
+#define NULL_SIZE 1
+
+static inline char mstrReqType(size_t string_size);
+static inline int mstrHdrSize(char type);
+static inline int mstrSumMetaLen(mstrKind *k, mstrFlags flags);
+static inline size_t mstrAllocLen(const mstr s, struct mstrKind *kind);
+
+/*** mstr API ***/
+
+/* Create mstr without any metadata attached, based on string 'initStr'.
+ * - If initStr equals NULL, then only allocation will be made.
+ * - string of mstr is always null-terminated.
+ */
+mstr mstrNew(const char *initStr, size_t lenStr) {
+    unsigned char *pInfo; /* pointer to mstr info field */
+    void *sh;
+    mstr s;
+    char type = mstrReqType(lenStr);
+    int mstrHdr = mstrHdrSize(type);
+
+    assert(lenStr + mstrHdr + 1 > lenStr); /* Catch size_t overflow */
+
+    sh = s_malloc(mstrHdr + lenStr + NULL_SIZE);
+
+    if (sh == NULL) return NULL;
+
+    s = (char*)sh + mstrHdr;
+    pInfo = ((unsigned char*)s) - 1;
+
+    switch(type) {
+        case MSTR_TYPE_5: {
+            *pInfo = CREATE_MSTR_INFO(lenStr, 0 /*ismeta*/, type);
+            break;
+        }
+        case MSTR_TYPE_8: {
+            MSTR_HDR_VAR(8,s);
```

Comment:
does this macro(mstrhdr##T *sh) will redefine with `void *sh`?

Reply:
Not sure if i understand the question. Here it will preprocessed to:
```
struct mstrhdr8 *sh = (void*)((s)-(sizeof(struct mstrhdr8)));
```

Reply:
but above it already have a `sh`.
```
    void *sh;  <- here
    mstr s;
    char type = mstrReqType(lenStr);
    int mstrHdr = mstrHdrSize(type);
```

Reply:
I followed the pattern of `sds`. Not the best, and yet.